### PR TITLE
Add some more matchers

### PR DESCRIPTION
**Pull request by msabramo**
- `PredFunc`
- `HasPrefix`
- `HasSuffix`
- `Regexp`

Cc: @seanisom, @mayanand, @balshetzer

### CODE REVIEW

**Comment by msabramo** [comment]
Bump

**Comment by msabramo** [comment]
Rebased and fixed merge conflict.

File: gomock/matchers.go [hunk 1]
```diff
@@ -139,3 +190,52 @@ func Not(x interface{}) Matcher {
 func AssignableToTypeOf(x interface{}) Matcher {
 	return assignableToTypeOfMatcher{reflect.TypeOf(x)}
 }
+
+// PredFunc is a Matcher that matches if the one argument function parameter
+// returns true for the given input
```
Comment (poy) [inline gomock/matchers.go]: nit: Can we make this commend doc a sentence with a period to match the other matchers?

File: gomock/matchers.go [hunk 2]
```diff
@@ -139,3 +190,52 @@ func Not(x interface{}) Matcher {
...
+//
+// Example usage:
+//
+// 		mock.EXPECT().
+// 			Handle(gomock.PredFunc(func(x interface{}) bool { return x.(int) > 5 })).
```
Comment (poy) [inline gomock/matchers.go]: nit: This is a little tricky to read all in one line. WDYTA:

```diff
diff --git a/gomock/matchers.go b/gomock/matchers.go
index db19cd6..88ce69b 100644
--- a/gomock/matchers.go
+++ b/gomock/matchers.go
@@ -197,7 +197,9 @@ func AssignableToTypeOf(x interface{}) Matcher {
 // Example usage:
 //
 //             mock.EXPECT().
-//                     Handle(gomock.PredFunc(func(x interface{}) bool { return x.(int) > 5 })).
+//                     Handle(gomock.PredFunc(func(x interface{}) bool {
+//                             return x.(int) > 5
+//                     })).
 //                     Return(true)
 //
```

File: gomock/matchers.go [hunk 3]
```diff
@@ -139,3 +190,52 @@ func Not(x interface{}) Matcher {
...
+// 			Return(true)
+//
+func PredFunc(f func(interface{}) bool, name string) Matcher {
+	return predFuncMatcher{predFunc: f, name: name}
+}
+
+// HasPrefix is a Matcher that matches if the input has the arg as a prefix
```
Comment (poy) [inline gomock/matchers.go]: nit: Can we make this commend doc a sentence with a period to match the other matchers?

File: gomock/matchers.go [hunk 4]
```diff
@@ -139,3 +190,52 @@ func Not(x interface{}) Matcher {
...
+//
+// 		mock.EXPECT().
+// 			Handle(gomock.HasPrefix("abc")).
+// 			Return(true)
+//
+func HasPrefix(prefix string) Matcher {
+	return hasPrefixMatcher{prefix: prefix}
+}
+
+// HasSuffix is a Matcher that matches if the input has the arg as a suffix
```
Comment (poy) [inline gomock/matchers.go]: nit: Can we make this commend doc a sentence with a period to match the other matchers?

File: gomock/matchers.go [hunk 5]
```diff
@@ -139,3 +190,52 @@ func Not(x interface{}) Matcher {
...
+//
+// 		mock.EXPECT().
+// 			Handle(gomock.HasSuffix(".exe")).
+// 			Return(true)
+//
+func HasSuffix(suffix string) Matcher {
+	return hasSuffixMatcher{suffix: suffix}
+}
+
+// Regexp is a Matcher that matches if the input matches the regular expression string arg
```
Comment (poy) [inline gomock/matchers.go]: nit: Can we make this commend doc a sentence with a period to match the other matchers?

File: gomock/matchers.go [hunk 6]
```diff
@@ -139,3 +190,52 @@ func Not(x interface{}) Matcher {
...
+//
+// Example usage:
+//
+// 		mock.EXPECT().
+// 			Handle(gomock.Regexp(".exe$")).
```
Comment (poy) [inline gomock/matchers.go]: super nit: Can you throw a backslash (`\`) in front of the period (`.`)? I'm assuming you are demonstrating a test that looks for a windows executable extension (`.exe`), but the regexp would actually match against anything ending in `exe`.

I feel silly for bringing it up, but all the same.

**Comment by codyoss** [comment]
@msabramo still interested in this PR?

**Comment by codyoss** [comment]
Closing due to inactivity. Please feel free to reopen if you are still interested in these changes.

**Pull request closed without merging**